server/datastore/mysql/software.go

```diff
@@ -2882,301 +3113,601 @@ last_vpp_install AS (
+		}
+
+		var softwareOnlySelfServiceClause string
+		var vppOnlySelfServiceClause string
+		if opts.SelfServiceOnly {
+			softwareOnlySelfServiceClause = ` AND software_installers.self_service = 1 `
+			if opts.IsMDMEnrolled {
+				vppOnlySelfServiceClause = ` OR ( vpp_apps_teams.self_service = 1 AND true ) `
+			} else {
+				vppOnlySelfServiceClause = ` OR ( vpp_apps_teams.self_service = 1 AND false ) `
+			}
+		}
 
-		// promote the package name and version to the proper destination fields
-		if hs.PackageName != nil {
-			var version string
-			if hs.PackageVersion != nil {
-				version = *hs.PackageVersion
+		var cveMetaFilter string
+		var hasCVEFilters bool
+		var cveMatchClause string
+		var cveNamedArgs []interface{}
+		var cveMatchArgs []interface{}
+		if opts.KnownExploit {
+			cveMetaFilter += "\nAND cve_meta.cisa_known_exploit = :known_exploit"
+			hasCVEFilters = true
+		}
+		if opts.MinimumCVSS > 0 {
+			cveMetaFilter += "\nAND cve_meta.cvss_score >= :min_cvss"
+			hasCVEFilters = true
+		}
+		if opts.MaximumCVSS > 0 {
+			cveMetaFilter += "\nAND cve_meta.cvss_score <= :max_cvss"
+			hasCVEFilters = true
+		}
+		if hasCVEFilters {
+			cveMetaFilter, cveNamedArgs, err = sqlx.Named(cveMetaFilter, namedArgs)
+			if err != nil {
+				return nil, nil, ctxerr.Wrap(ctx, err, "build named query for cve filters")
 			}
-			var platform string
-			if hs.PackagePlatform != nil {
-				platform = *hs.PackagePlatform
+		}
+		if opts.ListOptions.MatchQuery != "" {
+			cveMatchClause, cveMatchArgs = searchLike(cveMatchClause, cveMatchArgs, opts.ListOptions.MatchQuery, "software_cve.cve")
+		}
+
+		var softwareVulnerableJoin string
+		if len(softwareTitleIds) > 0 {
+			if opts.VulnerableOnly || opts.ListOptions.MatchQuery != "" {
+				softwareVulnerableJoin += " AND ( "
+				if !opts.VulnerableOnly && opts.ListOptions.MatchQuery != "" {
+					softwareVulnerableJoin += `
+						NOT EXISTS (
+							SELECT 1
+							FROM 
+								software_cve
+							WHERE
+								software_cve.software_id = software.id
+						) OR
+					`
+				}
+
+				softwareVulnerableJoin += `
+				EXISTS (
+					SELECT 1
+					FROM
+						software_cve
+				`
+				cveMetaJoin := "\n INNER JOIN cve_meta ON software_cve.cve = cve_meta.cve"
+
+				// Only join CVE table if there are filters
+				if hasCVEFilters {
+					softwareVulnerableJoin += cveMetaJoin
+				}
+				softwareVulnerableJoin += `
+					WHERE
+						software_cve.software_id = software.id
+				`
+				softwareVulnerableJoin += cveMetaFilter
+				softwareVulnerableJoin += "\n" + cveMatchClause
+				softwareVulnerableJoin += "\n))"
 			}
-			hs.SoftwarePackage = &fleet.SoftwarePackageOrApp{
-				Name:        *hs.PackageName,
-				Version:     version,
-				Platform:    platform,
-				SelfService: hs.PackageSelfService,
+
+			installedSoftwareJoinsCondition := ""
+			if len(softwareIDs) > 0 {
+				installedSoftwareJoinsCondition = `AND software.id IN (?)`
```

Comment:
Feels like this could be its own query, only when there are software IDs, only for those software IDs, and when we get results back we merge any found vulns in Go-land. Per previous discussions, a piece of software is scanned for vulns iff it's in inventory on the host, and vulnerabilities reported when software is both installed and available for install (or pending install/uninstall) are for whatever is in inventory, not the version being installed or available for install.

Reply:
https://github.com/fleetdm/fleet/issues/28293